### PR TITLE
LookupAccountName

### PR DESCRIPTION
**Pull request by splatteredbits**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
+# PureInvoke PowerShell Module Changelog
+
+## 1.0.0
+
+* Created `Invoke-AdvapiLookupAccountName` to call the Advapi `LookupAccountName` function.
+* Created `Invoke-AdvapiLookupAccountSid` to call the Advapi `LookupAccountName` function.
```
Comment (decoyjoe) [inline CHANGELOG.md]: Typo: Should be `LookupAccountSid` function.

File: PureInvoke/PureInvoke.psd1 [hunk 2]
```diff
@@ -76,6 +76,9 @@
 
     # Functions to export from this module. Only list public function here.
     FunctionsToExport = @(
+        'Invoke-AdvapiLookupAccountName',
+        'Invoke-AdvapiLookupAccountSid',
+        'Invoke-KernelFindFileName'
```
Comment (decoyjoe) [inline PureInvoke/PureInvoke.psd1]: No module prefix? 🤔

File: PureInvoke/src/advapi32.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+using System.Runtime.InteropServices;
+using System.Security.Principal;
+using System.Text;
+
+namespace PureInvoke
+{
+	public static class Advapi32
+	{
+		// ReSharper disable InconsistentNaming
+		[DllImport("advapi32", CharSet=CharSet.Auto, SetLastError=true)]
```
Comment (decoyjoe) [inline PureInvoke/src/advapi32.cs]: Missing the `.dll` extension.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    Demonstrates how to call this function by passing a username to the `AccountName` parameter.
+    #>
+    [CmdletBinding()]
+    param(
+        # The name of the system.
+        [String] $SystemName,
+
+        # The account name to lookup.
+        [Parameter(Mandatory)]
+        [String] $AccountName
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1]: List Mandatory parameters first?

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    [byte[]] $sid = [byte[]]::New(0);
+
+    # cb = count of bytes
+    [UInt32] $cbSid = 0;
+    [Text.StringBuilder] $domainName = [Text.StringBuilder]::New()
+    # cch = count of chars
+    [UInt32] $cchDomainName = $domainName.Capacity;
+    [PureInvoke.AdvApi32+SidNameUse] $sidNameUse = [PureInvoke.AdvApi32+SidNameUse]::Unknown;
+
+    [PureInvoke.ErrorCode]$errCode = [PureInvoke.ErrorCode]::Ok
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1]: Nit: Add a space between the type cast and the variable.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    $result = [PureInvoke.AdvApi32]::LookupAccountName($SystemName, $AccountName,
+                                                       $sid, [ref] $cbSid,
+                                                       $domainName, [ref] $cchDomainName,
+                                                       [ref]$sidNameUse)
+    $errCode = [Marshal]::GetLastWin32Error()
+    if ($result)
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1]: Should this be `(-not $result)` ?

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    {
+
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1]: Errant newline.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        # The account name to lookup.
+        [Parameter(Mandatory)]
+        [String] $AccountName
+    )
+
+    Set-StrictMode -Version 'Latest'
+    Use-CallerPreference -Cmdlet $PSCmdlet -Session $ExecutionContext.SessionState
+
+    $result = [pscustomobject]@{
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1]: Does this assignment service any purpose? It's re-assigned on line 52 before this assignment is ever used.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountSid.ps1 [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    Invoke-AdvapiLookupAccountSid -Sid $sid
+
+    Demonstrates how to call this function by passing a sid to the `Sid` parameter.
+    #>
+    [CmdletBinding()]
+    param(
+        [String] $SystemName,
+
+        [Parameter(Mandatory)]
+        [byte[]] $Sid
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvapiLookupAccountSid.ps1]: List Mandatory parameters first?

File: PureInvoke/Functions/Invoke-KernelFindFileName.ps1 [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+    Set-StrictMode -Version 'Latest'
+    Use-CallerPreference -Cmdlet $PSCmdlet -Session $ExecutionContext.SessionState
+
+    [PureInvoke.ErrorCode] $errCode = [PureInvoke.ErrorCode]::Ok
+
+    # Loop over and collect all hard links as their full paths.
+    [IntPtr]$findHandle = [IntPtr]::Zero
+
+    # Must be MAX_PATH otherwise calls to FindFirstFileName@ and FindNextFileNameW throw exceptions that crash
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-KernelFindFileName.ps1]: Typo: `@` should be `W`

Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-KernelFindFileName.ps1]: I'm not sure what code this comment is referring to?

File: PureInvoke/src/netapi32.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
+using System;
+using System.ComponentModel;
+using System.Runtime.InteropServices;
+
+namespace PureInvoke
+{
+	public static class NetApi32
```
Comment (decoyjoe) [inline PureInvoke/src/netapi32.cs]: I didn't see where this was used?

File: README.md [hunk 12]
```diff
@@ -12,15 +12,15 @@ The "PInvoke" module...
...
 To install privately:
 
 ```powershell
-Save-Module -Name 'PInvoke' -Path '.'
-Import-Module -Name '.\PInvoke'
+Save-Module -Name 'PureInvoke' -Path '.'
+Import-Module -Name '.\PureInvoke'
 ```
 
 # Commands
```
Comment (decoyjoe) [inline README.md]: Document the available module commands.

**Review by decoyjoe — CHANGES_REQUESTED** [review]

File: PureInvoke/PureInvoke.psd1 (hunk 2 above)
Comment (splatteredbits) [reply PureInvoke/PureInvoke.psd1]: The pattern will be for all the functions in the module to be `Invoke-LibFunction` where `Lib` is the library name, without the `32` and `Function` is the name of the function in the library. I thought about using a prefix, but figured the naming convention would suffice and already causes very long function names.

Any ideas for a prefix? `PI`? Skip?

File: PureInvoke/src/advapi32.cs (hunk 3 above)
Comment (splatteredbits) [reply PureInvoke/src/advapi32.cs]: This method actually not used so removed.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 (hunk 6 above)
Comment (splatteredbits) [reply PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1:55]: No `0x0` is success in Win32 land.

Comment (splatteredbits) [reply PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1]: Yeah. This function wasn't working at all. Fixed and added some tests.

File: PureInvoke/PureInvoke.psd1 (hunk 2 above)
Comment (decoyjoe) [reply PureInvoke/PureInvoke.psd1]: My initial thought was `PI` too. Another idea was `Interop` but that's long and doesn't match the module name. No other good ideas outside of that.

I don't really have strong feelings about it, so I'll let you decide, prefix or not.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 (hunk 4 above)
Comment (splatteredbits) [reply PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1]: Moved.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 (hunk 5 above)
Comment (splatteredbits) [reply PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1]: Fixed.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 (hunk 7 above)
Comment (splatteredbits) [reply PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1]: Removed.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 (hunk 8 above)
Comment (splatteredbits) [reply PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1]: Fixed.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountSid.ps1 (hunk 9 above)
Comment (splatteredbits) [reply PureInvoke/Functions/Invoke-AdvapiLookupAccountSid.ps1]: Moved.

File: PureInvoke/Functions/Invoke-KernelFindFileName.ps1 (hunk 10 above)
Comment (splatteredbits) [reply PureInvoke/Functions/Invoke-KernelFindFileName.ps1]: Removed comment as it doesn't apply to anything.

Comment (splatteredbits) [reply PureInvoke/Functions/Invoke-KernelFindFileName.ps1]: Removed.

File: PureInvoke/src/netapi32.cs (hunk 11 above)
Comment (splatteredbits) [reply PureInvoke/src/netapi32.cs]: Removed.

File: README.md (hunk 12 above)
Comment (splatteredbits) [reply README.md]: Done.

File: CHANGELOG.md [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+* Created `Invoke-AdvApiLsaAddAccountRights` function to call the advapi32.dll library's `LsaAddAccountRights` function.
+* Created `Invoke-AdvApiLsaClose` function to call the advapi32.dll library's `LsaClose` function.
+* Created `Invoke-AdvApiLsaEnumerateAccountRights` function to call the advapi32.dll library's
+  `LsaEnumerateAccountRights` function.
+* Created `Invoke-AdvApiLsaFreeMemory` function to call the advapi32.dll library's `LsaFreeMemory` function.
+* Created `Invoke-AdvApiLsaNtStatusToWinError` function to call the advapi32.dll library's `LsaNtStatusToWinError`
+  function.
+* Created `Invoke-AdvApiLsaOpenPolicy` function to call the advapi32.dll library's `LsaOpenPolicy` function.
+* Created `Invoke-AdvApiLsaRemoveAccountRights` function to call the advapi32.dll library's `LsaRemoveAccountRights`
+  function.
```
Comment (decoyjoe) [inline CHANGELOG.md:17]: In the future can we put additional feature changes in a separate PR? All these new additions make reviewing the original PR more difficult.

File: PureInvoke/Functions/Invoke-AdvApiLsaRemoveAccountRights.ps1 [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    <#
+    .SYNOPSIS
+    Calls the advapi32.dll library's `LsaRemoveAccountRights` method which removes rights/privileges for an account.
+
+    .DESCRIPTION
+    The `Invoke-AdvApiLsaRemoveAccountRights` function calls the advapi32.dll library's `LsaRemoveAccountRights` method
+    which removes rights/privileges for an account. Pass the LSA policy handle to the `PolicyHandle` parameter. Pass the
+    security identifier for the account to the `Sid` parameter. To remove *all* of the account's rights, use the `All`
+    switch. Otherwise, pass the specific rights to remove to the `Privilege` parameter. If the removal succeeds, the
+    function returns `$true`, otherwise it returns $false and writes an error.
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvApiLsaRemoveAccountRights.ps1]: "`$true`" is surrounded in backticks but "$false" isn't.

File: PureInvoke/Functions/Invoke-AdvApiLsaRemoveAccountRights.ps1 [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+    .EXAMPLE
+    Invoke-AdvApiLsaRemoveAccountRights -PolicyHandle $handle -Sid $sid -All
+
+    Demonstrates how to remove all of an account's privileges.
+
+    .EXAMPLE
+    Invoke-AdvApiLsaRemoveAccountRights -PolicyHandle $handle -Sid $sid -Privilege 'SeBatchLogonRight'
+
+    Demonstrates how to remove a specific account's privilege.
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvApiLsaRemoveAccountRights.ps1]: I think you meant:

> Demonstrates how to remove a specific privilege from an account.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    Demonstrates how to call this function by passing a username to the `AccountName` parameter.
+    #>
+    [CmdletBinding()]
+    param(
+        # The account name to lookup.
+        [Parameter(Mandatory)]
+        [String] $AccountName,
+
+        # The name of the system.
+        [String] $SystemName
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1]: `$SystemName` should be `$ComputerName` for consistency with other functions in this module.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    # cb = count of bytes
+    [UInt32] $cbSid = 0;
+    [StringBuilder] $sbDomainName = [StringBuilder]::New()
+    # cch = count of chars
+    [UInt32] $cchDomainName = $sbDomainName.Capacity;
+    [SidNameUse] $sidNameUse = [SidNameUse]::Unknown;
+
+    $result = [AdvApi32]::LookupAccountName($SystemName, $AccountName, $sid, [ref] $cbSid, $sbDomainName,
+                                            [ref] $cchDomainName, [ref]$sidNameUse)
+    $errCode = [Marshal]::GetLastWin32Error()
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1:47]: Should this statement be nested within the `(-not $result)` conditional?

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountSid.ps1 [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    Invoke-AdvApiLookupAccountSid -Sid $sid
+
+    Demonstrates how to call this function by passing a sid to the `Sid` parameter.
+    #>
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory)]
+        [byte[]] $Sid,
+
+        [String] $SystemName
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvapiLookupAccountSid.ps1]: Missing parameter docs.

Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvapiLookupAccountSid.ps1]: `$SystemName` should be `$ComputerName` for consistency with other functions in this module.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountSid.ps1 [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+    [UInt32] $cchName = $name.Capacity;
+
+    [StringBuilder] $domainName = [StringBuilder]::New()
+    [UInt32] $cchDomainName = $domainName.Capacity;
+
+    [SidNameUse] $sidNameUse = [SidNameUse]::Unknown;
+
+    $result = [AdvApi32]::LookupAccountSid($SystemName, $sid, $name, [ref] $cchName, $domainName, [ref] $cchDomainName,
+                                           [ref] $sidNameUse)
+    $errCode = [Marshal]::GetLastWin32Error()
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-AdvapiLookupAccountSid.ps1:48]: Should this statement be nested within the `(-not $result)` condition?

File: PureInvoke/Functions/Invoke-KernelGetVolumePathName.ps1 [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+        [Parameter(Mandatory)]
+        [String] $Path
+    )
+
+    Set-StrictMode -Version 'Latest'
+    Use-CallerPreference -Cmdlet $PSCmdlet -Session $ExecutionContext.SessionState
+
+    # Need to call GetFullPathName to get the size of the necessary buffer.
+    $sbPath = [Text.StringBuilder]::New([PureInvoke.Kernel32]::MAX_PATH)
+    $cchPath = [UInt32]$sbPath.Capacity; # in/out character-count variable for the WinAPI calls.
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-KernelGetVolumePathName.ps1]: Errant semicolon.

File: PureInvoke/Functions/Invoke-KernelGetVolumePathName.ps1 [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+    param(
+        # The path whose volume mount point to get.
+        [Parameter(Mandatory)]
+        [String] $Path
+    )
+
+    Set-StrictMode -Version 'Latest'
+    Use-CallerPreference -Cmdlet $PSCmdlet -Session $ExecutionContext.SessionState
+
+    # Need to call GetFullPathName to get the size of the necessary buffer.
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-KernelGetVolumePathName.ps1]: I don't think this comment applies to anything?

File: PureInvoke/Functions/Invoke-KernelGetVolumePathName.ps1 [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+    $sbPath = [Text.StringBuilder]::New([PureInvoke.Kernel32]::MAX_PATH)
+    $cchPath = [UInt32]$sbPath.Capacity; # in/out character-count variable for the WinAPI calls.
+    $result = [PureInvoke.Kernel32]::GetVolumePathName($Path, $sbPath, $cchPath)
+    # Get the volume (drive) part of the target file's full path (e.g., @"C:\")
```
Comment (decoyjoe) [inline PureInvoke/Functions/Invoke-KernelGetVolumePathName.ps1]: I don't think this comment applies to anything?

File: README.md [hunk 23]
```diff
@@ -1,26 +1,51 @@
...
+* `LookupAccountSid`: `Invoke-AdvApiLookupAccountSid`
+* `LsaAddAccountRights`: `Invoke-AdvApiLsaAddAccountRights`
+* `LsaClose`: `Invoke-AdvApiLsaClose`
+* `LsaEnumerateAccountRights`: `Invoke-AdvApiLsaEnumerateAccountRights`
+* `LsaFreeMemory`: `Invoke-AdvApiLsaFreeMemory`
+* `LsaNtStatusToWinError`: `Invoke-AdvApiLsaNtStatusToWinError`
+* `LsaOpenPolicy`: `Invoke-AdvApiLsaOpenPolicy`
+* `LsaRemoveAccountRights`: `Invoke-AdvApiLsaRemoveAccountRights`
+
+### Froom kernel32.dll
```
Comment (decoyjoe) [inline README.md]: Typo: `Froom`

**Review by decoyjoe — APPROVED** [review]

File: CHANGELOG.md (hunk 13 above)
Comment (splatteredbits) [reply CHANGELOG.md:17]: Yes. I didn't realize when I started work that this PR already existed.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 (hunk 16 above)
Comment (splatteredbits) [reply PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1]: Made this change across the entire module.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 (hunk 17 above)
Comment (splatteredbits) [reply PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1:47]: When debugging PowerShell, something in the debugger can much with the last win32 error, so it's good practice to retrieve the last error directly after the call that sets it.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountSid.ps1 (hunk 19 above)
Comment (splatteredbits) [reply PureInvoke/Functions/Invoke-AdvapiLookupAccountSid.ps1:48]: Ditto. Best to capture the last error before something else sets/changes it.

File: PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1 (hunk 17 above)
Comment (decoyjoe) [reply PureInvoke/Functions/Invoke-AdvapiLookupAccountName.ps1:47]: That's a good point.

**Pull request merged**